Alertinator.php

```diff
@@ -50,26 +51,106 @@ public function __construct($config) {
          } catch (Exception $e) {
             // If there's an error in your check functions, you damn better
             // know about it.
             $message = "Internal failure in check:\n" . $e->getMessage();
             $this->alertGroups(
-               new AlertinatorWarningException($message),
-               $alerteeGroups
+             new AlertinatorWarningException($message),
+             $alerteeGroups
             );
             // Rethrow so your standard exception handler also gets it.
             throw $e;
          }
       }
    }
 
-   protected function alertGroups($exception, $alerteeGroups) {
+   /**
+    * Threshold notification: determine if an all-clear alert should be sent,
+    * and if so, send it and reset the logger.
+   */
+   private function notifyClear(string $check, int $alertAfter, int $clearAfter, array $alerteeGroups) {
+      $log = $this->logger->readAlerts($check);
+      krsort($log);
+
+      // We only get here if there was at least 1 failure, but 1 failure may not
+      // exceed the alertAfter threshold. If the check succeeds without
+      // reaching the alert threshold, reset the log silently.
+      // Note: this will break terribly for bouncing errors, e.g. pass-> fail->
+      //       pass-> fail-> pass-> fail. Account for this.
+      if (count($log) < $alertAfter) {
+         $this->logger->resetAlerts($check);
+      } else if (count($log) >= $clearAfter) {
+         $clears = 0;
+         // Only notify a clear if the check passes > $clearAfter in a row.
+         foreach ($log as $alert) {
+            if (!$alert['status']) {
+               break;
+            }
+            $clears++;
+         }
+         if ($clears >= $clearAfter) {
+            $last = end($log);
+            $message = "The alert '$check' was cleared at " . date(DATE_RFC2822, $last['ts']) . ".";
+            // TODO: It is impossible here to know what exception level should
+            // be sent. In absence of this information we have to just blast it
+            // to everyone. Once https://github.com/iFixit/alertinator/issues/3
+            // is implemented, exception level should be available.
+            $e = new AlertinatorCriticalException($message);
+            $this->alertGroups($e, $alerteeGroups);
+            $this->logger->resetAlerts($check);
+         }
+      }
+   }
+
+   /**
+    * Threshold notification: determine if a failure alert should be sent, and
+    * if so, send it.
+   */
+   private function notifyFailure(string $check, int $alertAfter,
+    iterable $alerteeGroups, Exception $e, int $remindEvery) {
+      $log = $this->logger->readAlerts($check);
+      $fails = 0;
+
+      foreach ($log as $alert) {
+         if ($alert['status']) {
+            break;
+         }
+         $fails++;
+      }
+
+      $reminderTriggered = $fails > $alertAfter
+       && ($fails - $alertAfter) % $remindEvery == 0;
+
+      if ($fails === $alertAfter || $reminderTriggered) {
```

Comment:
This is a whole bunch of condensed logic.

Might be easier to read if you broke it up and named things
```
$critNumFails = $fails === $alertAfter;
$pastCritialNumFails = $fails > $alertAfter;
$sendThisInternval = (($fails - $alertAfter) % $remindEvery) == 0;

if ($critNumFails || ($pastCritNumFails && $sendThisInterval))
```

Reply:
good idea. readability > brevity.